src/performance_stats.cpp

```diff
@@ -0,0 +1,106 @@
+/*****************************************************************************
+ *
+ * This file is part of Mapnik (c++ mapping toolkit)
+ *
+ * Copyright (C) 2017 Artem Pavlenko
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ *****************************************************************************/
+
+#include <mapnik/performance_stats.hpp>
+
+namespace mapnik {
+
+template class MAPNIK_DECL singleton<timer_stats, CreateStatic>;
+
+void timer_stats::add(std::string const& metric_name, double cpu_elapsed, double wall_clock_elapsed)
+{
+#ifdef MAPNIK_THREADSAFE
+    std::lock_guard<std::mutex> lock(metrics_mutex_);
```

Comment:
The lock guard will also be measured as it will require time to be constructed and deconstructed. Therefore, we are possibly measuring time to acquire the mutex and release it.